src/ray/util/subreaper.h

```diff
@@ -0,0 +1,126 @@
+// Copyright 2024 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#ifdef __linux__
+#include <signal.h>
+#include <sys/types.h>
+
+#include <boost/asio.hpp>
+#include <vector>
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/synchronization/mutex.h"
+#include "ray/util/macros.h"
+#endif
+
+namespace ray {
+
+#ifdef __linux__
+// Utility functions to enable subreaper functionality in Linux.
+//
+// In Ray, raylet creates core_worker processes, which may create grandchild processes.
+// If core_worker exits normally it tries to kill all its children recursively. However
+// if core_worker is killed by a signal, its children leak and still occupy resources.
```

Comment:
```suggestion
// if core_worker is crashed unexpectedly, its children leak and still occupy resources.
```

If signal is sigterm, it doesn't leak